design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google-finance
+  namespace: finance
+spec:
+  # note that this is a vertex, so there is no virtualhost key
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific, however as this is a vertex, only prefixes that match the prefix
+  # that delegated to this object.
+  routes:
+  # each route in this object must start with /finance as this was the prefix which
+  # delegated to this object.
+  - match: /finance
+    service:
+    - name: finance-app
+      port: 9999
+  - match: /finance/static
+    service:
+    - name: finance-static-content
+      port: 8080
+  - match: /finance/partners
+    # delegate the /finance/partners prefix to the partners namespace
+    delegate:
+      name: finance-partners
+      namespace: partners
+```
+
+## Delegation rules
+
+The delegation rules applied are as follows
+
+1. If an `IngressRoute` object contains a `spec.virtualhost` key it is considered a root.
+2. If an `IngressRoute` object does not contain `spec.virtualhost` key is considered a vertex.
+3. A vertex is reachable if a delegation to it exists in another `IngressRoute` object.
+4. Vertices which are not reachable are considered orphened. Orphened vertices have no effect on the running configuration.
+
+## Validation rules
+
+The validation rules applied in this design are as follows
+Some of these rules may be relaxed in future designs.
+
+1. If a validation error is encountered, the entire object is rejected. Partial application of the valid portions of the configuration is _not_ attempted.
+2. The prefix of the delegate route in the parent must be match the routes in the child.
+  ```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: root
+  namespace: root
+spec:
+  virtualhost:
+    fqdn: heptio.com
+    aliases:
+      - www.heptio.com
+  routes:
+  - match: /static
+    delegate:
+      name: child
+      namespace: child
+---
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: child 
+  namespace: child
+spec:
+  routes:
+  # valid
+  - match: /static/css
+  # invalid
+  - match: /css
+```    
+
+## Authorisation
+
+It is important to highlight that both root and vertex IngressRoute objects are of the same type.
+This is a departure from other designs which treat the permission to create a VirtualHost type object and a Route type object as separate.
+The DAG design treats the delegation from one IngressRoute to another as permission to create routes.
+
+## Reporting status
+
+The presence of semantically valid object is not proof that it will be used.
+This is a break from the kubernetes model whereby an object which is valid will generally be acted on by controllers.
```

Comment:
With ingress classes (an annotation, for sure) a controller will ignore an ingress for a class it doesn't understand.  Not exactly apples to apples but not totally different either.

Reply:
@jbeda are there other precedents here? ingress class is the only one I can think. I think in a different world this would have been managed with labels or something more akin to taints/tolerations.

Reply:
I think i answered this above but the GitHub review UI is fucking with me.